### PR TITLE
Theodore/proofnarrowing

### PR DESCRIPTION
**Pull request by theodorebugnet**
Add proof narrowing capabilities to the tree: given a range proof, supply two sets of leafs (contiguous with the edges of the original range), to generate a proof for the remaining sub-range.

This feature is useful for generating proofs of individual blobs in a Celestia namespace using the full-namespace proof returned by the `GetSharesByNamespace` RPC call.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -21,7 +21,7 @@ nmt-rs = { path = ".", features = ["borsh", "serde"] }
 borsh = { version = "1" }
 serde_json = "1.0.96"
 postcard = { version = "1.0.4", features = ["use-std"] }
-tendermint = { version = "0.35.0" }
```
Comment (preston-evans98) [inline Cargo.toml:24]: (Question): Why do we downgrade `tendermint` here?

File: src/lib.rs [hunk 2]
```diff
@@ -587,6 +587,78 @@ mod tests {
...
+        let root = tree.root();
+        for i in 1..=n {
+            for j in 0..=i {
+                let proof = tree.build_range_proof(j..i);
+                let leaf_hashes: Vec<_> = tree.leaves()[j..i]
+                    .iter()
+                    .map(|l| l.hash().clone())
+                    .collect();
+                for k in (j + 1)..i {
+                    for l in j..k {
```
Comment (preston-evans98) [inline src/lib.rs:603]: nit: Let's add a brief comment justifying why the end of the range (`k`) is excluded here

File: src/nmt_proof.rs [hunk 3]
```diff
@@ -98,6 +98,54 @@ where
         )
     }
 
+    /// Narrows the proof range: uses an existing proof to create
+    /// a new proof for a subrange of the original proof's range
+    pub fn narrow_range<L: AsRef<[u8]>>(
```
Comment (preston-evans98) [inline src/nmt_proof.rs:103]: Let's add some documentation explaining how to find the correct values for `left_extra_raw_leaves` and `right_extra_raw_leaves`

File: src/nmt_proof.rs [hunk 4]
```diff
@@ -98,6 +98,54 @@ where
...
+        let mut tree = NamespaceMerkleTree::<NoopDb, M, NS_ID_SIZE>::with_hasher(
+            M::with_ignore_max_ns(self.ignores_max_ns()),
+        );
+
+        let proof = tree.inner.narrow_range_proof(
+            &left_extra_hashes,
+            self.start_idx() as usize..(self.range_len() - new_leaf_len),
+            &right_extra_hashes,
+            &mut self.siblings(),
+            self.start_idx() as usize,
```
Comment (preston-evans98) [inline src/nmt_proof.rs:140]: let's add comments justifying why this range is the correct one

Comment (preston-evans98) [reply src/nmt_proof.rs:140]: In fact, this looks like it may be a bug. The argument we're passing is called `new_leaf_range`, but itseems like we don't account for which side of the tree the leaves are coming from in the calculation `self.start_idx() as usize..(self.range_len() - new_leaf_len)` - is that right? If so, maybe we need to rename the argument

File: src/simple_merkle/tree.rs [hunk 5]
```diff
@@ -352,6 +364,158 @@ where
...
+
+        // If the left subtree doesn't overlap with the new leaf, get its root and add it to the proof
+        if new_leaf_range.start >= (split_point) {
+            let sibling = self.partial_tree_subroot_inner(
+                working_range.start..split_point,
+                left_extra_leaves,
+                current_proof,
+                leaves_start_idx,
+            )?;
+            out.push(sibling.clone());
```
Comment (preston-evans98) [inline src/simple_merkle/tree.rs:389]: (Question): Do we need to `clone` here?

File: src/simple_merkle/tree.rs [hunk 6]
```diff
@@ -352,6 +364,158 @@ where
...
+        } else {
+            let subtrie_size = split_point - working_range.start;
+            assert!(subtrie_size > 0); // sanity check
+            if subtrie_size == 1 {
+                // If it's a leaf, ensure we have it, and do nothing
+                let index = working_range.start;
+                assert!(new_leaf_range.contains(&index));
```
Comment (preston-evans98) [inline src/simple_merkle/tree.rs:396]: Are we confident that these assertions can't be triggered by a malformed input? If so, I'm happy to keep them - but let's document why it's not possible to cause a crash. Otherwise let's change to returning an error

File: src/simple_merkle/tree.rs [hunk 7]
```diff
@@ -352,6 +364,158 @@ where
...
+        &self,
+        working_range: Range<usize>,
+        left_extra_leaves: &[M::Output],
+        new_leaf_range: &Range<usize>,
+        right_extra_leaves: &[M::Output],
+        current_proof: &mut &[M::Output],
+        leaves_start_idx: usize,
+        out: &mut Vec<M::Output>,
+    ) -> Result<(), RangeProofError> {
+        assert!(working_range.len() > 1); // sanity check
```
Comment (preston-evans98) [inline src/simple_merkle/tree.rs:378]: (Question) Should we handle this case instead of asserting?

File: src/nmt_proof.rs [hunk 8]
```diff
@@ -98,6 +98,54 @@ where
...
+        right_extra_raw_leaves: &[L],
+        leaf_namespace: NamespaceId<NS_ID_SIZE>,
+    ) -> Result<Self, RangeProofError> {
+        if self.is_of_absence() {
+            return Err(RangeProofError::MalformedProof(
+                "Cannot narrow the range of an absence proof",
+            ));
+        }
+
+        let new_leaf_len = left_extra_raw_leaves.len() + right_extra_raw_leaves.len();
```
Comment (preston-evans98) [inline src/nmt_proof.rs:115]: nit: It's not obvious from the name whether the `new_leaf_len` refers to the number of leaves *included* in the new proof, or to the number of leaves that were newly provided to facilitate proof generation.

File: src/simple_merkle/tree.rs [hunk 9]
```diff
@@ -352,6 +364,158 @@ where
...
+
+        Ok(())
+    }
+
+    /// To be used during the narrowing operation
+    /// Calculates a new subroot to be part of the narrowed proof,
+    /// in an area covered by the old proof and new leaves
+    fn partial_tree_subroot_inner(
+        &self,
+        subtrie_range: Range<usize>,
```
Comment (preston-evans98) [inline src/simple_merkle/tree.rs:453]: nit: It's not clear from the naming if this range is relative to the entire NMT, relative to the original proof, or relative to the `extra_leaves`

File: Cargo.toml (hunk 1 above)
Comment (theodorebugnet) [reply Cargo.toml:24]: I upgraded it - since it was out of date and I didn't see any breaking changes. Happy to revert if we prefer the dependencies to be more frozen

File: src/simple_merkle/tree.rs (hunk 5 above)
Comment (theodorebugnet) [reply src/simple_merkle/tree.rs:389]: Yep, we don't consume the proof (as it can be used to generate multiple sub-proofs) so instead we just clone out the individual siblings when we need them

File: src/nmt_proof.rs (hunk 4 above)
Comment (theodorebugnet) [reply src/nmt_proof.rs:140]: That is an _excellent_ point, and upon further inspection the test I added does not actually test this implementation in `nmt_proof` - only the one in `simple_merkle` directly. I'll fix and add a test for it. Thanks for the catch!

File: src/simple_merkle/tree.rs (hunk 7 above)
Comment (theodorebugnet) [reply src/simple_merkle/tree.rs:378]: Great point. It should only ever come up on a tree of size 1, but if so the operation should be a no-op, not an error.
Actually I need to see why this didn't come up in the test, since I think it does in fact run on a tree of size 1

File: src/simple_merkle/tree.rs (hunk 6 above)
Comment (theodorebugnet) [reply src/simple_merkle/tree.rs:396]: I'll double-check. Most of them shouldn't, e.g. once we get a range of size 1 that's a base case so we should never encounter a size 0. I'll add some extra comments.
This one I'll actually think about more, because the `new_leaf_range` _should_ just be naturally calculated base on the lengths of the inputs provided (and so would only fail if there's an internal bug in handling the ranges/indices), but I'll trace through it again to make sure you can't craft an input where this is false

File: src/simple_merkle/tree.rs [hunk 10]
```diff
@@ -352,6 +364,158 @@ where
...
+
+        // We are operating on a full subtree. So the base cases are:
+        // [####] - the added leaves are covering the entire range; use them to calculate the subroot
+        // [____] - there are no added leaves in the range; there is an existing proof node for this subtree
+        // In all other cases, we split as normal and recurse on both subtrees.
+
+        let leaves_end_idx = leaves_start_idx + extra_leaves.len();
+        if leaves_start_idx <= subtrie_range.start && leaves_end_idx >= subtrie_range.end {
+            local_subroot_from_leaves(extra_leaves, leaves_start_idx, subtrie_range, &self.hasher)
+        } else if leaves_start_idx >= subtrie_range.end || leaves_end_idx <= subtrie_range.start {
```
Comment (preston-evans98) [inline src/simple_merkle/tree.rs:495]: A quick comment on these conditions explaining how the checks relate to the cases you explain above would be super helpful. It's not easy to keep track of (for example) if `leaves_end_idx` is absolute or relative.

File: src/nmt_proof.rs (hunk 8 above)
Comment (theodorebugnet) [reply src/nmt_proof.rs:115]: I'll make a pass on the naming, you're right that it's not all great

File: src/simple_merkle/tree.rs [hunk 11]
```diff
@@ -352,6 +364,158 @@ where
...
+                extra_leaves,
+                current_proof,
+                leaves_start_idx,
+            )?;
+            let right = self.partial_tree_subroot_inner(
+                split_point..subtrie_range.end,
+                extra_leaves,
+                current_proof,
+                leaves_start_idx,
+            )?;
```
Comment (preston-evans98) [inline src/simple_merkle/tree.rs:514]: nit: We do a lot of recursion where only one argument changes. We could make it more obvious which arguments are changing and which ones are constant by wrapping it in a closure, or maybe bundling a few of the arguments into a struct.

**Review by preston-evans98 — COMMENTED** [review]
Based on the new test (which is very nice, btw!) this seems to work when the inputs are well-constructed. Unfortunately, it's pretty difficult to be confident about the correctness at a logical level given the current state of the code. Since this is pretty involved logic, I think it's probably worth investing in even more detailed explanations (probably including diagrams) in the core functions

File: Cargo.toml (hunk 1 above)
Comment (preston-evans98) [reply Cargo.toml:24]: Oops, yeah I read this backwards. Tendermint versioning is pretty... non-standard, so let's keep the version until we hear from Lumina that they want a newer one

**Comment by theodorebugnet** [comment]
Actually your comment about the ranges is spot-on - the current test was missing a whole bunch of edge cases due to being non-inclusive. Whoops!
I'm fixing that now, and _then_ it will probably be fine to assert that "it seems to work".

Happy to add some more comments and explanations, especially now that I've slept on it and can look at it from a slight distance!

**Comment by theodorebugnet** [comment]
By the way, the codecov.io token seems dead, so the badge in the README isn't updating anymore

File: src/simple_merkle/tree.rs [hunk 12]
```diff
@@ -352,6 +378,197 @@ where
...
     }
 
+    /// Helper for the proof narrowing operation.
+    ///
+    /// # Arguments:
+    /// - params: the immutable data used during recursion
+    /// - working_range: The range of leaf indices, relative to the entire tree, being currently
+    ///   considered. Recursion starts with Range(0..tree_size).
+    ///   outside the desired new narrower range.
+    ///   the right edge of the old range.
```
Comment (preston-evans98) [inline src/simple_merkle/tree.rs:388]: nit: It looks like this comment may have gotten truncated

**Review by preston-evans98 — APPROVED** [review]
Thanks for making a pass on the naming and docs. It's *much* easier to understand this time around. Looks great, thanks @theodorebugnet!
